### PR TITLE
ZEPPELIN-1130 Make Livy create session retries configurable

### PR DESCRIPTION
**Pull request by r-kamath**
### What is this PR for?
Make Livy create session retires configurable

### What type of PR is it?
Improvement

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1130

### How should this be tested?
- Add/Edit `zeppelin.livy.create.session.retries` property in Livy interpreter settings
- Run a paragraph with `%livy sc.version`
- Verify the retries count in logs

### Questions:
* Does the licenses files need update? n/a
* Is there breaking changes for older versions? n/a
* Does this needs documentation? n/a

### CODE REVIEW

**Comment by prabhjyotsingh** [comment]
LGTM!

File: livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java [hunk 1]
```diff
@@ -83,9 +82,16 @@ public Integer createSession(InterpreterContext context, String kind) throws Exc
...
       if (!jsonMap.get("state").equals("idle")) {
-        Integer nosRetry = MAX_NOS_RETRY;
+        Integer retryCount = 120;
 
-        while (nosRetry >= 0) {
+        try {
+          retryCount = Integer.valueOf(
+              property.getProperty("zeppelin.livy.create.session.retries"));
+        } catch (Exception e) {
+          //nothing to do
```
Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: LOGGER.warning?

File: livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java [hunk 2]
```diff
@@ -83,9 +82,16 @@ public Integer createSession(InterpreterContext context, String kind) throws Exc
           }.getType());
       Integer sessionId = ((Double) jsonMap.get("id")).intValue();
       if (!jsonMap.get("state").equals("idle")) {
-        Integer nosRetry = MAX_NOS_RETRY;
+        Integer retryCount = 120;
```
Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: 120 times? that seems a bit extreme? :)

File: livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java [hunk 3]
```diff
@@ -108,9 +114,9 @@ public Integer createSession(InterpreterContext context, String kind) throws Exc
             LOGGER.error(String.format("Cannot start  %s.\n%s", kind, logs));
             throw new Exception(String.format("Cannot start  %s.\n%s", kind, logs));
           }
-          nosRetry--;
+          retryCount--;
         }
-        if (nosRetry <= 0) {
+        if (retryCount <= 0) {
           LOGGER.error("Error getting session for user within 60Sec.");
```
Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: if we have 120 x 1000 ms this message should instead say "120 sec"?

File: livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java (hunk 2 above)
Comment (r-kamath) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: @felixcheung what would be the ideal default retry count here? can we revert it to 60 ?

Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: I'm not sure actually, I just thought trying something for 120 times is a bit much :)

Comment (r-kamath) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: I agree 120 is a bit too much to retry!. Will let the user decide. Reverting it back to 60 as default

**Comment by r-kamath** [comment]
@felixcheung made some changes. please review

File: livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java [hunk 4]
```diff
@@ -117,7 +118,7 @@ public Integer createSession(InterpreterContext context, String kind) throws Exc
           retryCount--;
         }
         if (retryCount <= 0) {
-          LOGGER.error("Error getting session for user within 60Sec.");
+          LOGGER.error("Error getting session for user within the given number of retries.");
```
Comment (felixcheung) [inline livy/src/main/java/org/apache/zeppelin/livy/LivyHelper.java]: minor suggestion: instead of "given", how about "configured"?

**Comment by felixcheung** [comment]
looks good! minor suggestion.

**Comment by prabhjyotsingh** [comment]
Merging this if no more discussion.

**Pull request closed without merging**